### DISCUSSION TITLE
ARM64EC / ARM64X support for runtime and NativeAOT

### DISCUSSION THREAD

**Discussion by govert** [dotnet/runtime#78192] [Ideas]
[ARM64EC](https://learn.microsoft.com/en-us/windows/ARM/ARM64ec) is a new application binary interface (ABI) for apps running on ARM devices with Windows 11. An ARM64EC process can load and run x64 code, and thus can host the .NET runtime and libraries that are compiled to target x64. However, running the x64 version of .NET inside an ARM64EC process has a performance cost - this code runs under emulation and not at full speed and the various optimizations that have been made in .NET for ARM64 would not be applied.

It seems like most of the operating system binaries under Windows 11 for ARM are built as dual-mode [ARM64X ](https://learn.microsoft.com/en-us/windows/ARM/ARM64x-pe)binaries, so can be loaded into both ARM64 and ARM64EC processes. A similar requirement holds for .NET in order to achieve full performance on this platform. We would like a .NET process to run as an ARM64EC process so that it can be compatible with x64 native libraries, and we would like to be able to host the ARM version of the .NET runtime inside an ARM64EC process, instead of being forced to run the x64 version of the runtime under emulation, as is currently the case. Similarly, for NativeAOT compilation we would like to have the option to target ARM64EC or ARM64X binary format.

**Comment by huoyaoyuan** [comment]
/cc @jkotas Is ARM64EC applicable to CLR?

I also wonder if ARM64 CLR can P/Invoke with x64 native libraries.

**Reply by MichalStrehovsky** [reply]
A recent discussion is here: https://github.com/dotnet/designs/pull/217#issuecomment-1861133275. We don't have plans for ARM64EC support for Native AOT.

**Comment by driver1998** [comment]
If you really need ARM64X (say you are building an in-process COM server), you can use ARM64X forwarder and point to x64 and ARM64 versions of the DLL respectively. 

https://learn.microsoft.com/en-us/windows/arm/arm64x-pe#arm64x-for-use-with-middleware-or-plugins

**Comment by trcrsired** [comment]
arm64ec is a COMPLETE MISTAKE by breaking intel APX

**Reply by tannergooding** [reply]
Can you elaborate on how you believe Arm64EC breaks the APX ISA extensions?
